lib/derivativo/conversion/office_helpers.rb

```diff
@@ -0,0 +1,106 @@
+        )
+      end
+
+      def self.soffice_binary_path_from_config_or_path
+        DERIVATIVO['soffice_binary_path'] || Derivativo::UserPathHelper.which('soffice')
+      end
+
+      def self.converted_file_size_merits_higher_compression_attempt?(file_size)
+        # TODO: At some point, maybe change PDF_HIGHER_COMPRESSION_CONVERSION_THRESHOLD here into a configuration option instead
+        # of a hard-coded value, though that will also change the logic in the
+        # compression_value_for_first_try_file_size method.
+        file_size > PDF_HIGHER_COMPRESSION_CONVERSION_THRESHOLD
+      end
+
+      def self.compression_value_for_first_try_file_size(size)
+        return 8 if size > 2000.megabytes
+        return 80 if size <= PDF_HIGHER_COMPRESSION_CONVERSION_THRESHOLD
+        # This formula seems to give decent results, but might need more adjustments in the future.
+        ((1 / Math.log(size.to_f / 1.megabyte, 10))**2 * 100).to_i
+      end
+
+      def self.office_convert_to_pdf_impl(src_file_path:, dst_file_path:, soffice_binary_path:, compression_integer: 80)
+        # Create a unique, temporary home dir for this office conversion process so we can run
+        # multiple conversion jobs independently. If two conversion processes use the same
+        # home dir, the first process will block the second one.
+        Derivativo::FileHelper.working_directory_temp_dir('office_temp_homedir') do |office_temp_homedir|
+          # Then write out an office settings file to set the desired PDF compression level.
+          create_office_settings_file(office_temp_homedir.path, compression_integer)
+
+          # Now we'll create a unique, temporary directory where we'll put our pre-renamed output file.
+          # The office conversion process always keeps the original name of the file and replaces
+          # the extension with 'pdf', so we'll need to predict the tmp outpath and move to the
+          # expected dst_file_path after we're done with the conversion.
+          Derivativo::FileHelper.working_directory_temp_dir('office_temp_outdir') do |office_temp_outdir|
+            # Run conversion
+            Open3.capture2(cmd_to_run = conversion_command(soffice_binary_path, src_file_path, office_temp_homedir.path, office_temp_outdir.path))
+
+            # The office conversion process always keeps the original name of the file and replaces
+            # the extension with 'pdf', so we'll need to predict the tmp outpath and move it after.
+            expected_conversion_outpath = File.join(office_temp_outdir, File.basename(src_file_path).sub(/(.+)\..+$/, '\1.pdf'))
+            Derivativo::FileHelper.block_until_file_exists(expected_conversion_outpath)
+
+            # Move the generated file to the correct out_path
+            if File.exist?(expected_conversion_outpath)
+              FileUtils.mv(expected_conversion_outpath, dst_file_path)
+              Derivativo::FileHelper.block_until_file_exists(dst_file_path)
```

Comment:
I'm assuming (and I may be wrong) that FileUtils.mv only returns once the mv is complete or failed, and is not executing inside a different thread, 
- Assuming we still want to check that the file was moved, I would suggest just a File.exist? call instead of Derivativo::FileHelper.block_until_file_exists
- The result of the check for file existence is not use inside this method, or in the calling method (office_convert_to_pdf)

Reply:
My understanding is that FileUtils.mv returns only once the move is complete or failed, yes.  The problem I'm working around here is that sometimes a network-mounted volume (when under read/write load) doesn't recognize the presence of a new file for a couple of seconds.  I'm doing the block here so that the caller of this method doesn't need to be aware of any special waiting procedure, and can assume that a file is ready to go when the method returns.